src/modules/simulation/gz_plugins/moving_platform_controller/MovingPlatformController.cpp

```diff
@@ -0,0 +1,325 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "MovingPlatformController.hpp"
+
+using namespace custom;
+
+// Register the plugin
+GZ_ADD_PLUGIN(
+	MovingPlatformController,
+	gz::sim::System,
+	gz::sim::ISystemPreUpdate,
+	gz::sim::ISystemConfigure
+)
+
+void MovingPlatformController::Configure(const gz::sim::Entity &entity,
+		const std::shared_ptr<const sdf::Element> &sdf,
+		gz::sim::EntityComponentManager &ecm,
+		gz::sim::EventManager &eventMgr)
+{
+	_entity = entity;
+	_model = gz::sim::Model(entity);
+
+	const std::string link_name = sdf->Get<std::string>("link_name");
+	_link_entity = _model.LinkByName(ecm, link_name);
+
+	if (!_link_entity) {
+		throw std::runtime_error("MovingPlatformController::Configure: Link \"" + link_name + "\" was not found. "
+					 "Please ensure that your model contains the corresponding link.");
+	}
+
+	_link = gz::sim::Link(_link_entity);
+
+	// Needed to report linear & angular velocity
+	_link.EnableVelocityChecks(ecm, true);
+
+	_startup_timer = gz::common::Timer();
+	_startup_timer.Start();
+
+	// Get velocity and orientation setpoints from env vars
+	{
+		const double vel_forward = ReadEnvVar("PX4_GZ_PLATFORM_VEL", 1.0);
+		const double heading_deg = ReadEnvVar("PX4_GZ_PLATFORM_HEADING_DEG", 0.0);
+		const double heading = GZ_DTOR(heading_deg);
+
+		_orientation_sp = gz::math::Quaterniond(0., 0., heading);
+
+		// Set initial orientation to match the heading
+		const auto optional_original_pose = _link.WorldPose(ecm);
+
+		if (optional_original_pose.has_value()) {
+			const gz::math::Pose3d original_pose = optional_original_pose.value();
+			const gz::math::Pose3d new_pose(original_pose.Pos(), _orientation_sp);
+			_model.SetWorldPoseCmd(ecm, new_pose);
+
+		} else {
+			gzwarn << "Unable to get initial platform pose. Heading will not be set" << std::endl;
+		}
+
+		// Velocity setpoint in world frame.
+		const gz::math::Vector3d _body_velocity_sp(vel_forward, 0., 0.);
+		_velocity_sp = _orientation_sp * _body_velocity_sp;
+	}
+
+	// Get gravity, model mass, platform height.
+	{
+		const auto world_entity = gz::sim::worldEntity(ecm);
+		const auto world = gz::sim::World(world_entity);
+		const auto gravity = world.Gravity(ecm);
+
+		if (gravity.has_value()) {
+			_gravity = gravity.value().Z();
```

Comment:
You can also disable gravity for this link?

Reply:
I think it is good enough to go in at this point, I just think there are ways to improve it